Fix StatisticsLogger Deadlock

DieHertz opened the pull request:
For a detailed description have a look at #1329.

Review and suggestions are welcome, in case I've missed something.

DieHertz commented:
Seems like in this variant it also causes a more obvious, but a deadlock :-(

DieHertz commented:
When `StatisticsLogger` is initialized from `subscribe` thread, `subs_mutex_` gets locked twice in the sequence `subscribe (topic) -> advertise (statistics)`. `subs_mutex_` could be made a `recursive_mutex` to allow locking it more than once.

Also, `StatisticsLogger` shouldn't be initialized for topics `/clock` and `/statistics` to avoid locking inside `ros::init`.

Another solution (which works in my tests, doesn't reproduce the original issue, but I can't be sure it's entirely correct yet) is to introduce a `advertise_subscribe_mutex_` into `TopicManager` and lock it inside `advertise/unadvertise/subscribe/unsubscribe`.

Which one should I go with? Need your advice.

DieHertz committed:
Solve StatisticsLogger deadlock by extracting initialization

DieHertz commented:
Decided to go with another approach, extracting `StatisticLogger::pub_` initialization to `TopicManager::subscribe`, but outside of `Subscription` creation.
That way I can no longer reproduce any of the issues I had nor other associated deadlock that were possible.

kev-the-dev commented on clients/roscpp/include/ros/statistics.h:
```diff
@@ -84,6 +87,7 @@ class ROSCPP_DECL StatisticsLogger
   double pub_frequency_;
 
   // publisher for statistics data
+  boost::atomic<bool> pub_initialized_;
```
Above comment is no longer accurate.

DieHertz replied:
Will swap the lines

kev-the-dev reviewed (commented):
EDIT: See the issue for the backtrace and code insights what show the deadlock scenario

I've been unable to reproduce this on my machine with the following program like the one you mentioned in the issue (I've run dozens of times). Still looking into it... Do you have a  specific program I can see this with/a rostest which produces this?:
```
#include <ros/ros.h>
#include <std_msgs/Int32.h>
#include <iostream>
#include <vector>

void c1(const std_msgs::Int32::ConstPtr& msg)
{
  std::cout << "C1" << std::endl;
}

void c2(const std_msgs::Int32::ConstPtr& msg)
{
  std::cout << "C2" << std::endl;
}

int main(int argc, char** argv) {
  ros::init(argc, argv, "statistics_logger_deadlock");
  ros::NodeHandle nh;
  
  std::vector<ros::Subscriber*> subs;
  // This call returns properly
  for (size_t i = 0; i < 100; ++i)
  {
    ros::Subscriber* sub = new ros::Subscriber();
    *sub = nh.subscribe<std_msgs::Int32>("/topic", 1, c1);
    ros::spinOnce();
    subs.push_back(sub);
  }

  std::cout << "Bout to spin" << std::endl;

  ros::spin();
}
```

DieHertz commented:
To reproduce it I have run a really CPU intensive computation task on the same machine.
Visual inspection of the code and a captures trace from 2 locked threads shows that the situation is indeed possible, albeit not too frequent.

I'll try to come up with a minimal project to reproduce this weekend though.

kev-the-dev commented:
No need, I can see it in the code now. I'll checkout your solution

kev-the-dev reviewed (approved):
I agree with this solution. Only downside is that some messages could be skipped if the callback is called before initializeStatisticsLogger() runs. Would appreciate comments above the new functions / variables describing what they do.

DieHertz commented:
I have reworked it a bit in order to simplify, looks like two separate `init*` functions for `StatisticsLogger` aren't necessary.
Also added comments where appropriate.

dirk-thomas commented:
Please retarget the latest default branch (currently `melodic-devel`) to be considered for review and merge.

dirk-thomas commented:
Closing due to no further activity.
